### PR TITLE
Wrap new synchronous logins API (from a-s) in an async version

### LINKED ISSUE #812: Unify FxaResult and SyncResult
See #807 and #811. I think the way forward here is to:

1. Put something that does the work of these in the base component, and move FxA to use it
2. Expose a minimally safe synchronous API from application-services.
3. Move the bulk of application-services's API into android-components, including all the async code, which will now use the result type from the base component.

### PR DESCRIPTION
**Pull request by thomcc**
This is the a-c version of https://github.com/mozilla/application-services/pull/383.

Fixes #812, by killing `SyncResult`.

r? @csadilek

### CODE REVIEW

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/AsyncLoginsStorage.kt [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+import kotlinx.coroutines.plus
+import mozilla.appservices.logins.DatabaseLoginsStorage
+import mozilla.appservices.logins.LoginsStorage
+import mozilla.appservices.logins.MemoryLoginsStorage
+
+/**
+ * This type contains the set of information required to successfully
+ * connect to the server and sync.
+ */
+typealias SyncUnlockInfo = mozilla.appservices.logins.SyncUnlockInfo
```
Comment (csadilek) [inline components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/AsyncLoginsStorage.kt]: 👍

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/AsyncLoginsStorage.kt [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    override fun add(login: ServerPassword): Deferred<String> {
+        return scope.async { wrapped.add(login) }
+    }
+
+    override fun update(login: ServerPassword): Deferred<Unit> {
+        return scope.async { wrapped.update(login) }
+    }
+
+    override fun close() {
+        wrapped.close()
```
Comment (csadilek) [inline components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/AsyncLoginsStorage.kt]: Should we call `job.cancel()` here?

File: samples/sync-logins/src/main/java/org/mozilla/samples/sync/logins/SyncLoginsClient.kt [hunk 3]
```diff
@@ -1,21 +1,27 @@
...
+import mozilla.components.service.sync.logins.AsyncLoginsStorageAdapter
 
 /**
  * A minimal login example.
  *
  * Note that this code likely has issues if accessed by multiple threads (especially around
  * initialization and lock/unlock).
  */
-class SyncLoginsClient(private val databasePath: String) {
+class SyncLoginsClient(private val databasePath: String) : AutoCloseable {
```
Comment (csadilek) [inline samples/sync-logins/src/main/java/org/mozilla/samples/sync/logins/SyncLoginsClient.kt]: `databasePath` doesn't need to be a val/property.

File: samples/sync-logins/src/main/java/org/mozilla/samples/sync/logins/SyncLoginsClient.kt [hunk 4]
```diff
@@ -28,57 +34,26 @@ class SyncLoginsClient(private val databasePath: String) {
...
-            if (wasLocked) {
-                // This shouldn't happen given how we've structured this example, but it's simple
-                // to handle, and worth handling in the case that some of this is copy-pasted into
-                // situations where this is possible.
-                SyncResult.fromValue(false)
-            } else {
-                store.lock().then { SyncResult.fromValue(true) }
-            }
-        }
+    suspend fun getLocalPasswordList(): List<ServerPassword> {
```
Comment (csadilek) [inline samples/sync-logins/src/main/java/org/mozilla/samples/sync/logins/SyncLoginsClient.kt]: This method could be `private`.

File: components/service/sync-logins/README.md [hunk 5]
```diff
@@ -106,141 +104,6 @@ HTML field name of the username, if known.
 
 HTML field name of the password, if known.
 
-## `LoginsStorage`
```
[the remark this answers is not in the document]
Comment (thomcc) [reply components/service/sync-logins/README.md]: Sure. I removed things that were duplicated compared to the doc comments, but I'm fine adding some back.

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/AsyncLoginsStorage.kt (hunk 2 above)
Comment (thomcc) [reply components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/AsyncLoginsStorage.kt]: Probably!

File: components/service/sync-logins/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (thomcc) [reply components/service/sync-logins/README.md]: I added a usage example, let me know if you think it's enough.

**Comment by sashei** [comment]
🎉 !!

**Pull request merged**